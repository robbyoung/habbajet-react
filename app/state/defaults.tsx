import {HabbajetEditor, PurchaseEditor} from '.';
import {habbajetColors} from '../colors';

const EMPTY_FIELD = {
    value: '',
    errorMessage: '',
};

export const DEFAULT_EDITOR_STATE: HabbajetEditor = {
    name: EMPTY_FIELD,
    value: EMPTY_FIELD,
    modifier: EMPTY_FIELD,
    slack: EMPTY_FIELD,
    color: habbajetColors[0],
    validated: false,
};

export const DEFAULT_PURCHASE_EDITOR_STATE: PurchaseEditor = {
    name: EMPTY_FIELD,
    cost: EMPTY_FIELD,
    tagId: '',
    validated: false,
};
