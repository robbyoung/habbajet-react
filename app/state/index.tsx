export enum HabitResult {
    Success,
    SlackSuccess,
    Failure,
}

export interface Habbajet {
    id: string;
    name: string;
    maxValue: number;
    currentValue: number;
    modifier: number;
    results: HabitResult[];
    color: string;
    date: string;
    toClaim: boolean;
    selected: boolean;
    currentStreak: number;
    bestStreak: number;
    oldStreaks: [number, number];
    totalSlack: number;
    remainingSlack: number;
}

export interface Purchase {
    id: string;
    name: string;
    cost: number;
    date: string;
    tagId: string;
}

export interface EditorField {
    value: string;
    errorMessage: string;
}

export interface HabbajetEditor {
    name: EditorField;
    value: EditorField;
    modifier: EditorField;
    slack: EditorField;
    color: string;
    validated: boolean;
}

export interface PurchaseEditor {
    name: EditorField;
    cost: EditorField;
    tagId: string;
    validated: boolean;
    id?: string;
}

export interface Tag {
    id: string;
    name: string;
    color: string;
}

export interface State {
    habbajets: Habbajet[];
    purchases: Purchase[];
    tags: Tag[];
    habbajetEditor: HabbajetEditor;
    purchaseEditor: PurchaseEditor;
    budget: number;
}
