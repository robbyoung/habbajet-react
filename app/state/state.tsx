export interface Habbajet {
    name: string;
    value: number;
    days: boolean[];
}

export interface Purchase {
    name: string;
    value: number;
}

export interface State {
    habits: Habbajet;
    purchages: Purchase;
    budget: string;
}
