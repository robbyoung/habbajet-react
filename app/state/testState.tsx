import {State, Habbajet, Purchase} from '.';

export function createTestState(
    numHabbajets: number,
    numPurchases: number,
    budget: number,
): State {
    const habbajets: Habbajet[] = [];
    for (let i = 0; i < numHabbajets; i++) {
        habbajets.push({
            name: `Habbajet ${i}`,
            maxValue: 100,
            currentValue: 100 / Math.pow(2, 7),
            factor: 2,
            successes: 0,
            color: '#8066C2',
            date: '2020-03-22T11:00:00.000Z',
            toClaim: false,
        });
    }

    const purchases: Purchase[] = [];
    for (let i = 0; i < numPurchases; i++) {
        purchases.push({
            name: `Purchase ${i}`,
            value: i,
        });
    }

    return {
        habbajets,
        purchases,
        budget,
    };
}
