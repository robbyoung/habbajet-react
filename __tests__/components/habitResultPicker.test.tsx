import React from 'react';
import renderer from 'react-test-renderer';
import HabitResultPicker from '../../app/components/habitResultPicker';
import {createTestState} from '../../app/state/testState';
import {render, fireEvent, wait} from '@testing-library/react-native';

jest.mock('@fortawesome/react-native-fontawesome', () => ({
    FontAwesomeIcon: '',
}));

describe('HabitResultPicker Component', () => {
    it('renders buttons and the specified day of the week', () => {
        const habbajet = createTestState(1, 0, 0).habbajets[0];
        habbajet.date = '2020-03-21T20:44:24.967Z';
        const component = renderer.create(
            <HabitResultPicker
                habbajet={habbajet}
                onFailure={() => undefined}
                onSuccess={() => undefined}
            />,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('collapses for dates in the future', () => {
        const habbajet = createTestState(1, 0, 0).habbajets[0];
        habbajet.date = '3000-01-01T00:00:00.000Z';
        const component = renderer.create(
            <HabitResultPicker
                habbajet={habbajet}
                onFailure={() => undefined}
                onSuccess={() => undefined}
            />,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('collapses for habbajets that can be claimed', () => {
        const habbajet = createTestState(1, 0, 0).habbajets[0];
        habbajet.toClaim = true;
        const component = renderer.create(
            <HabitResultPicker
                habbajet={habbajet}
                onFailure={() => undefined}
                onSuccess={() => undefined}
            />,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('will run the onSuccess callback if the success button is pressed', async () => {
        const habbajet = createTestState(1, 0, 0).habbajets[0];
        const onSuccess = jest.fn();
        const {getByTestId} = render(
            <HabitResultPicker
                habbajet={habbajet}
                onFailure={() => undefined}
                onSuccess={onSuccess}
            />,
        );

        const button = getByTestId('button-success');
        fireEvent.press(button);
        await wait(() => expect(onSuccess).toBeCalled());
    });

    it('will run the onFailure callback if the failure button is pressed', async () => {
        const habbajet = createTestState(1, 0, 0).habbajets[0];
        const onFailure = jest.fn();
        const {getByTestId} = render(
            <HabitResultPicker
                habbajet={habbajet}
                onFailure={onFailure}
                onSuccess={() => undefined}
            />,
        );

        const button = getByTestId('button-failure');
        fireEvent.press(button);
        await wait(() => expect(onFailure).toBeCalled());
    });
});
