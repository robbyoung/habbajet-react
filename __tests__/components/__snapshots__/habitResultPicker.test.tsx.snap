// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitResultPicker Component collapses for dates in the future 1`] = `<View />`;

exports[`HabitResultPicker Component collapses for habbajets that can be claimed 1`] = `<View />`;

exports[`HabitResultPicker Component renders buttons and the specified day of the week 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "column",
        "padding": 10,
      },
      Object {
        "backgroundColor": "#8066C2",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#fff",
          "fontFamily": "Abel",
          "fontSize": 30,
          "textAlign": "center",
          "width": "100%",
        },
      ]
    }
  >
    Sunday, March 22nd
  </Text>
  <View
    style={
      Array [
        Object {
          "alignContent": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <
        color="#fff"
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f00c",
              "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
            ],
            "iconName": "check",
            "prefix": "fas",
          }
        }
        size={100}
      />
    </View>
    <View
      accessible={true}
      focusable={false}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <
        color="#fff"
        icon={
          Object {
            "icon": Array [
              352,
              512,
              Array [],
              "f00d",
              "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
            ],
            "iconName": "times",
            "prefix": "fas",
          }
        }
        size={100}
      />
    </View>
  </View>
</View>
`;
